Cpp_test/lib/ipc-shm-receive.cpp

```diff
@@ -0,0 +1,120 @@
+#include "ipc-shm-receive.h"
+
+IPCShmReceive::~IPCShmReceive()
+{
+	if (this->shm_ptr != nullptr)
+	{
+		int ret1 = pthread_cond_destroy(&(this->shm_ptr->cond));
+		if (ret1 != 0)
+		{
+			std::cerr << "Error at pthread_cond_destroy(): " << strerror(ret1) << std::endl;
+		}
+
+		pthread_mutex_trylock(&(this->shm_ptr->mutex));
+
+		unlock_mutex();
+
+		int ret3 = pthread_mutexattr_destroy(&(this->mutex_attr));
+		if (ret3 != 0)
+		{
+			std::cerr << "Error at pthread_mutexattr_destroy(): " << strerror(ret3) << std::endl;
+		}
+
+		int ret4 = pthread_mutex_destroy(&(this->shm_ptr->mutex));
+		if (ret4 != 0)
+		{
+			std::cerr << "Error at pthread_mutex_destroy(): " << strerror(ret4) << std::endl;
+		}
+	}
+}
+
+void IPCShmReceive::init()
+{
+	this->file_handler.open_file();
+	this->open_shm();
+	this->map_shm();
+	if (this->shmd > 0)
+	{
+		close(this->shmd);
+	}
+}
+
+void IPCShmReceive::transfer()
+{
+	std::cout << "Receiving..." << std::endl;
+	unsigned long total_received_bytes{0};
+
+	while (!this->shm_ptr->is_end && this->shm_ptr->is_init)
+	{
+		lock_mutex();
+		while (this->shm_ptr->data_version_received == this->shm_ptr->data_version)
+		{
+			if (pthread_cond_wait(&(this->shm_ptr->cond), &(this->shm_ptr->mutex)) != 0)
+			{
+				unlock_mutex();
+				throw std::runtime_error(std::string("ERROR: pthread_cond_wait() send: ") + strerror(errno));
```

Comment:
`Upon successful completion, a value of zero shall be returned; otherwise, an error number shall be returned to indicate the error`
So errno is not useful here.